mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java

```diff
@@ -148,48 +157,63 @@ private GroupAwareQueryManagerImpl(final ScheduledExecutorServiceWrapper schedul
    * and executes the sources in order to receives data streams.
    * Before the queries are executed, it stores the avro  dag into disk.
    * We can regenerate the queries from the stored avro dag.
-   * @param tuple a pair of the query id and the avro dag
+   * @param avroDag the avro dag
    * @return submission result
    */
   @Override
-  public QueryControlResult create(final Tuple<String, AvroDag> tuple) {
+  public QueryControlResult create(final AvroDag avroDag) {
+    return createWithCheckpointedStates(avroDag, null);
+  }
+
+  /**
+   * Start a submitted query with the checkpointed states.
+   * @param avroDag the avro dag
+   * @param checkpointedStates the checkpointed states
+   * @return
+   */
+  @Override
+  public QueryControlResult createWithCheckpointedStates(final AvroDag avroDag,
+                                                         final QueryCheckpoint checkpointedStates) {
     final QueryControlResult queryControlResult = new QueryControlResult();
-    queryControlResult.setQueryId(tuple.getKey());
+    final String queryId = avroDag.getQueryId();
+    queryControlResult.setQueryId(queryId);
     try {
       // Create the submitted query
-      // 1) Saves the avr dag to the PlanStore and
-      // converts the avro dag to the logical and execution dag
-      planStore.saveAvroDag(tuple);
-      final String queryId = tuple.getKey();
-
+      // 1) Saves the avr dag and converts the avro dag to the logical and execution dag
+      checkpointManager.storeQuery(avroDag);
       // Update app information
-      final String appId = tuple.getValue().getAppId();
+      final String appId = avroDag.getAppId();
 
       if (LOG.isLoggable(Level.FINE)) {
         LOG.log(Level.FINE, "Create Query [aid: {0}, qid: {2}]",
             new Object[]{appId, queryId});
       }
 
       if (!applicationMap.containsKey(appId)) {
-        createApplication(appId, tuple.getValue().getJarPaths());
+        createApplication(appId, avroDag.getJarPaths());
       }
 
       final ApplicationInfo applicationInfo = applicationMap.get(appId);
-      final DAG<ConfigVertex, MISTEdge> configDag = configDagGenerator.generate(tuple.getValue());
+      final DAG<ConfigVertex, MISTEdge> configDag;
+      if (checkpointedStates == null) {
+        configDag = configDagGenerator.generate(avroDag);
+      } else {
+        configDag = configDagGenerator.generateWithCheckpointedStates(avroDag, checkpointedStates);
+      }
       // Waiting for group information being added
       while (applicationInfo.getGroups().isEmpty()) {
         Thread.sleep(100);
       }
       final Query query = createAndStartQuery(queryId, applicationInfo, configDag);
```

Comment:
This is probably another issue, but all recovered queries should be created before they are actually started.

Reply:
Could you elaborate on why we should create all the queries before we actually start them?

Reply:
Stateful queries can be merged.
Suppose queries A and B share stateful operators.
If A is recovered and processes some events, it may change the state of an operator that is also included in B, which has yet not been recovered.

Reply:
OK. In this case, we can make barrier for recovered queries. Could you make an issue for it?

Reply:
I've created #1039.

Reply:
Thanks!